### PR TITLE
FIX for issue #1537 related to #1524 + Support for non-FQN in web-API.

### PR DESCRIPTION
**Pull request by phoenix128**
This is a more robust fix for issue https://github.com/magento-engcom/msi/issues/1524 .
Also see: https://github.com/magento-engcom/msi/issues/1537 .

### Description
Other than fixing https://github.com/magento-engcom/msi/issues/1524 and https://github.com/magento-engcom/msi/issues/1537 , this commit also provides a Magento2 support for **non-FQN classes** in interfaces.

Before this commit you were not able to import class names in interfaces used for web-API.

For example, the following interface was not working with the non-FQN name `MyClass` imported through the `use` alias definition.
```
namespace ...;

use My\Module\MyClass;

interface SomeInterface
{
...
/**
 * @param MyClass $param
 * ...
 */
 public function getMyMethod()
 ...
...
}
```

### Fixed Issues (if relevant)
1. magento-engcom/msi#1537: Provide robust fix for nested object instantiation via Web API
2. magento-engcom/msi#1524: POST V1/inventory/source-selection-algorithm-result throws exception

### Contribution checklist
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [X] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: lib/internal/Magento/Framework/Reflection/TypeProcessor.php [hunk 1]
```diff
@@ -521,14 +522,78 @@ public function getParamType(ParameterReflection $param)
...
+            $classTypeReflection = new ClassReflection($typeName);
+            return $typeName . $arraySuffix;
+        } catch (\ReflectionException $e) {
+            unset($e);
+        }
+
+        // Resolve fully qualified name
+        $sourceFileName = $sourceClass->getDeclaringFile();
+        $source = $sourceFileName->getContents();
+        $parser = new TokenParser($source);
```
Comment (LeeSaferite) [inline lib/internal/Magento/Framework/Reflection/TypeProcessor.php]: Why go this route vs. using `$sourceFileName->getUses()` instead?

Comment (phoenix128) [reply lib/internal/Magento/Framework/Reflection/TypeProcessor.php]: You are right, I did not know about `getUses`. I am refactoring now. Many thanks.

**Comment by phoenix128** [comment]
Cross reference: https://github.com/magento/magento2/pull/17424 .

**Comment by maghamed** [comment]
see https://github.com/magento/magento2/pull/16881

**Comment by phoenix128** [comment]
Update:
The branch has green tests on https://github.com/magento/magento2/pull/17424 .
Not passing here due to frozen code files.

**Comment by ishakhsuvarov** [comment]
Fixed and merged in the scope of https://github.com/magento/magento2/pull/17424

Thanks @phoenix128

**Pull request closed without merging**